app-example/src/main/java/com/badoo/ribs/example/photo_details/feature/PhotoDetailsFeature.kt

```diff
@@ -39,45 +42,82 @@ internal class PhotoDetailsFeature(
         private val photoDetailsDataSource: PhotoDetailsDataSource,
-        private val photoId: String
+        private val photoId: String,
+        private val authDataSource: AuthDataSource
     ) : Actor<State, Wish, Effect> {
         override fun invoke(state: State, wish: Wish): Observable<Effect> =
             when (wish) {
                 Wish.LoadPhoto -> photoDetailsDataSource.getPhoto(photoId)
                     .map<Effect> { Effect.PhotoLoaded(it) }
                     .toObservable()
                     .startWith(Effect.LoadingStarted)
-                Wish.LikePhoto -> photoDetailsDataSource.likePhoto(photoId)
-                    .toObservable<Effect>()
-                    .startWith(Effect.LikeSent)
+                Wish.LikePhoto -> likeOrDislikePhoto(state)
             }
+
+        private fun likeOrDislikePhoto(state: State): Observable<Effect> =
+            if (state is State.Loaded && authDataSource.getState() is Authenticated) {
+                if (state.detail.likedByUser) {
+                    unlikePhoto()
+                } else {
+                    likePhoto()
+                }
+            } else Observable.empty()
+
+        private fun likePhoto(): Observable<Effect> {
+            return photoDetailsDataSource.likePhoto(photoId)
+                .toObservable<Effect>()
+                .startWith(Effect.LikeSent)
+        }
+
+        private fun unlikePhoto(): Observable<Effect> {
+            return photoDetailsDataSource.likePhoto(photoId)
+                .toObservable<Effect>()
+                .startWith(Effect.UnlikeSent)
+        }
     }
 
     class ReducerImpl : Reducer<State, Effect> {
         override fun invoke(state: State, effect: Effect): State =
             when (effect) {
                 is Effect.LoadingStarted -> State.Loading
                 is Effect.PhotoLoaded -> State.Loaded(effect.detail)
-                is Effect.LikeSent -> if (state is State.Loaded) {
-                    state.copy(detail = state.detail.copy(likedByUser = true))
-                } else {
-                    state
-                }
+                is Effect.LikeSent -> likeState(state, true)
+                is Effect.UnlikeSent -> likeState(state, false)
+            }
+
+        private fun likeState(state: State, isLiked: Boolean): State {
+            return if (state is State.Loaded) {
+                state.copy(detail = state.detail.copy(likedByUser = isLiked))
+            } else {
+                state
             }
+        }
     }
 
-    class NewsPublisherImpl : NewsPublisher<Wish, Effect, State, News> {
+    class NewsPublisherImpl(
+        private val authDataSource: AuthDataSource
+    ) : NewsPublisher<Wish, Effect, State, News> {
         override fun invoke(wish: Wish, effect: Effect, state: State): News? =
-            null
+            when (wish) {
+                is Wish.LikePhoto -> {
+                    if (authDataSource.getState() !is Authenticated) {
```

Comment:
I definitely wouldn't include this logic here, but in Actor (this logic is part of execution chain), and emit corresponding Effect.

Reply:
Nvm, I was commenting on outdated file.